packages/flutter/lib/src/material/menu_bar.dart

```diff
@@ -0,0 +1,3235 @@
+  Timer? hoverTimer;
+  Timer? clickBanTimer;
+  bool clickBan = false;
+  late FocusNode focusNode;
+
+  bool get _isAnOpenMenu {
+    return menu != null && controller!.isAnOpenMenu(menu!);
+  }
+
+  @override
+  void initState() {
+    super.initState();
+    focusNode = FocusNode(debugLabel: widget.label);
+    focusNode.addListener(_markDirty);
+  }
+
+  @override
+  void dispose() {
+    focusNode.removeListener(_markDirty);
+    focusNode.dispose();
+    super.dispose();
+  }
+
+  @override
+  void didChangeDependencies() {
+    _updateChangedState();
+    super.didChangeDependencies();
+  }
+
+  @override
+  void didUpdateWidget(MenuBarMenu oldWidget) {
+    _updateChangedState();
+    super.didUpdateWidget(oldWidget);
+  }
+
+  bool get enabled => controller!.enabled && widget.menus.isNotEmpty;
+
+  @override
+  Widget build(BuildContext context) {
+    return MenuBarItem._forMenu(
+      label: widget.label,
+      labelWidget: widget.labelWidget,
+      onSelected: enabled ? _maybeToggleShowMenu : null,
+      onHover: enabled ? _handleMenuHover : null,
+      focusNode: focusNode,
+      leadingIcon: widget.leadingIcon,
+      trailingIcon: widget.trailingIcon,
+      semanticLabel: widget.semanticLabel,
+      backgroundColor: widget.buttonBackgroundColor,
+      foregroundColor: widget.buttonForegroundColor,
+      overlayColor: widget.buttonOverlayColor,
+      textStyle: widget.buttonTextStyle,
+      padding: widget.buttonPadding,
+      shape: widget.buttonShape,
+      menuPadding: widget.padding ?? MenuBarTheme.of(context).menuPadding ?? _TokenDefaultsM3(context).menuPadding,
+    );
+  }
+
+  void _markDirty() {
+    setState(() {});
+  }
+
+  void _updateChangedState() {
+    menu = _MenuNodeWrapper.of(context);
+    controller = MenuBarController.of(context) as _MenuBarController;
+  }
+
+  // Shows the submenu if there is one, and it wasn't visible. Hides the menu if
+  // it was already visible.
+  void _maybeToggleShowMenu() {
+    if (clickBan) {
+      // If we just opened the menu because the user is hovering, then ignore
+      // clicks for a bit.
+      return;
+    }
+
+    if (_isAnOpenMenu) {
+      controller!.close(menu!);
+    } else {
+      controller!.openMenu = menu;
+    }
+  }
+
+  // Called when the pointer is hovering over the menu button.
+  void _handleMenuHover(bool hovering) {
+    // Cancel any click ban in place if hover changes.
+    clickBanTimer?.cancel();
+    clickBanTimer = null;
+    clickBan = false;
+
+    // Don't open the top level menu bar buttons on hover unless something else
+    // is already open. This means that the user has to first open the menu bar
+    // before hovering allows them to traverse it.
+    if (menu!.parent == controller!.root && controller!.openMenu == null) {
+      return;
+    }
+
+    if (hovering && !_isAnOpenMenu) {
+      controller!.openMenu = menu;
+      // If we just opened the menu because the user is hovering, then just
```

Comment:
Well, I think it's the sort of thing you have to try out to see what I mean. Linux (and maybe Windows, I didn't try) does this. You can't close macOS submenus with a click (and so maybe we shouldn't either on macOS, Hmmm.), so it doesn't do it.

Basically, it's one of those muscle memory things that we're all used to, but don't realize it.  Try clicking a top level menu, and then sliding down quickly and clicking on a submenu to open it.  Notice that when you click (if you do it fast enough), it will ignore the click and leave the menu open. If you do it slowly, the click will close the menu.